hugegraph-core/src/main/java/com/baidu/hugegraph/traversal/algorithm/PathsTraverser.java

```diff
@@ -61,19 +62,21 @@ public PathSet paths(Id sourceV, Directions sourceDir,
             return paths;
         }
 
+        boolean findCycle = false;
         Id labelId = this.getEdgeLabelId(label);
         Traverser traverser = new Traverser(sourceV, targetV, labelId,
                                             degree, capacity, limit);
+        // We should stop early if find cycle or reach limit
         while (true) {
-            if (--depth < 0 || traverser.reachLimit()) {
+            if (--depth < 0 || findCycle ||traverser.reachLimit()) {
                 break;
             }
-            traverser.forward(sourceDir);
+            findCycle = traverser.forward(targetV, sourceDir);
 
-            if (--depth < 0 || traverser.reachLimit()) {
+            if (--depth < 0 || findCycle || traverser.reachLimit()) {
                 break;
             }
-            traverser.backward(targetDir);
+            findCycle = traverser.backward(sourceV, targetDir);
```

Comment:
finish traverse immediately if find cycle？

Reply:
split to another pr